File: BSvsZP-Common/Messages/Reply.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Common;

namespace Messages
{
    abstract public class Reply : Message
    {
        #region Private Properties
        private static Int16 ClassId { get { return (Int16)MESSAGE_CLASS_IDS.Reply; } }
        #endregion

        #region Public Properties
        public PossibleTypes ReplyType { get; set; }
        public PossibleStatus Status { get; set; }
        public string Note {get; set;}
        public static new int MinimumEncodingLength
        {
            get
            {
                return 4                // Object header
                       + 1              // ReplyType
                       + 1              // Status
                       + 2;             // Note
            }
        }

        public enum PossibleTypes
        {
            AckNak = 1,
            ReadyReply = 2,
            ResourceReply = 3,
            ConfigurationReply = 4,
            PlayingFieldReply = 5,
            AgentListReply = 6,
            StatusReply = 7
        }

        public enum PossibleStatus
        {
            Success = 1,
            Failure = 2
        }

        #endregion

        #region Constructors and Factory Methods
        /// <summary>
        /// Default message constructor, used by Factory methods (i.e., the Create methods)
        /// </summary>
        protected Reply() { }

        /// <summary>
        /// Constructor used by all specializations, which in turn are used by
        /// senders of a message 
        /// </summary>
        /// <param name="type">Type of request that being created</param>
        /// <param name="status">Status of the ack/nak</status>
        /// <param name="note">error message or note</note>
        protected Reply(PossibleTypes type, PossibleStatus status, string note)
        {
            ReplyType = type;
            Status = status;
            Note = note;
        }

        /// <summary>
        /// Factor method to create a message from a byte list
        /// </summary>
        /// <param name="messageBytes">A byte list from which the message will be decoded</param>
        /// <returns>A new message of the right specialization</returns>
        new public static Reply Create(ByteList messageBytes)
        {
            Reply result = null;

            if (messageBytes == null || messageBytes.RemainingToRead < MinimumEncodingLength)
                throw new ApplicationException("Invalid message byte array");

            Int16 msgType = messageBytes.PeekInt16();
            switch (msgType)
            {
                case (Int16) MESSAGE_CLASS_IDS.AckNak:
                    result = AckNak.Create(messageBytes);
                    break;
                case (Int16) MESSAGE_CLASS_IDS.ReadyReply:
                    break;
                case (Int16) MESSAGE_CLASS_IDS.ResourceReply:
                    break;
                case (Int16) MESSAGE_CLASS_IDS.ConfigurationReply:
                    break;
                case (Int16) MESSAGE_CLASS_IDS.PlayingFieldReply:
                    break;
                case (Int16) MESSAGE_CLASS_IDS.AgentListReply:
                    break;
                case (Int16) MESSAGE_CLASS_IDS.StatusReply:
                    break;
                default:
                    throw new ApplicationException("Invalid Message Class Id");
            }

            return result;
        }
        #endregion

        #region Encoding and Decoding methods

        /// <summary>
        /// This method encodes
        /// </summary>
        /// <param name="bytes"></param>
        public override void Encode(ByteList bytes)
        {
            bytes.Add(ClassId);                              // Write out this class id first

            Int16 lengthPos = bytes.CurrentWritePosition;    // Get the current write position, so we
                                                                    // can write the length here later
            bytes.Add((Int16)0);                             // Write out a place holder for the length

            base.Encode(bytes);                              // Encode stuff from base class

            bytes.Add(Convert.ToByte(ReplyType));            // Write out a place holder for the length

            bytes.Add(Convert.ToByte(Status));               // Write out a place holder for the length

            if (Note == null) Note = string.Empty;
            bytes.Add(Note);

            Int16 length = Convert.ToInt16(bytes.CurrentWritePosition - lengthPos - 2);
            bytes.WriteInt16To(lengthPos, length);           // Write out the length of this object        
        }

        /// <summary>
        /// This method decodes a message from a byte list
        /// </summary>
        /// <param name="bytes"></param>
        public override void Decode(ByteList bytes)
        {
            Int16 objType = bytes.GetInt16();
            Int16 objLength = bytes.GetInt16();

            bytes.SetNewReadLimit(objLength);

            base.Decode(bytes);

            ReplyType = (PossibleTypes)Convert.ToInt32(bytes.GetByte());
            Status = (PossibleStatus)Convert.ToInt32(bytes.GetByte());
            Note = bytes.GetString();

            bytes.RestorePreviosReadLimit();
        }

        #endregion

    }
}

File: BSvsZP-Common -Java/Messages/Reply.java
package Messages;

import java.io.NotActiveException;
import java.net.UnknownHostException;

import org.omg.CORBA.portable.ApplicationException;

import Common.ByteList;
import Messages.Message.MESSAGE_CLASS_IDS;

public abstract class Reply extends Message {

    public enum PossibleTypes {

        AckNak(1),
        ReadyReply(2),
        ResourceReply(3),
        ConfigurationReply(4),
        PlayingFieldReply(5),
        AgentListReply(6),
        StatusReply(7);

        private int value;

        PossibleTypes(int value) {
            this.value = value;
        }

        public int getValue() {
            return value;
        }

        public static short getStringValueFromInt(int i) {
            for (PossibleTypes status : PossibleTypes.values()) {
                if (status.getValue() == i) {
                    return (short) status.value;
                }
            }

            throw new IllegalArgumentException("the given number doesn't match any Status.");
        }

        public static PossibleTypes convert(int value) {
            PossibleTypes temp = null;
            for (PossibleTypes dt : PossibleTypes.values()) {
                if (dt.value == value) {
                    temp = dt;
                }
            }
            return temp;

        }

    }

    public enum PossibleStatus {

        Success(1),
        Failure(2);

        private int value;

        PossibleStatus(int value) {
            this.value = value;
        }

        public int getValue() {
            return value;
        }

        public static short getStringValueFromInt(int i) {
            for (PossibleStatus status : PossibleStatus.values()) {
                if (status.getValue() == i) {
                    return (short) status.value;
                }
            }

            throw new IllegalArgumentException("the given number doesn't match any Status.");
        }

        public static PossibleStatus convert(int value) {
            PossibleStatus temp = null;
            for (PossibleStatus dt : PossibleStatus.values()) {
                if (dt.value == value) {
                    temp = dt;
                }
            }
            return temp;
        }
    }
    public PossibleTypes ReplyType;
    public PossibleStatus Status;
    public String Note;
    private static int MinimumEncodingLength;

    protected Reply() {
    }

    protected Reply(PossibleTypes type, PossibleStatus status, String note) {
        ReplyType = type;
        Status = status;
        Note = note;
    }

    public static Reply Create(ByteList messageBytes) throws Exception {
        Reply result = null;

        if (messageBytes == null || messageBytes.getRemainingToRead() < 6) {
            throw new ApplicationException("Invalid message byte array", null);
        }

        short msgType = messageBytes.PeekInt16();

        if (msgType == (short) MESSAGE_CLASS_IDS.AckNak.getValue()) 
            result = AckNak.Create(messageBytes);
         else if (msgType == (short) MESSAGE_CLASS_IDS.ReadyReply.getValue()) 
            result = ReadyReply.Create(messageBytes);
         else if (msgType == (short) MESSAGE_CLASS_IDS.ResourceReply.getValue()) 
            result = ResourceReply.Create(messageBytes);
         else if (msgType == (short) MESSAGE_CLASS_IDS.ConfigurationReply.getValue()) 
            result = ResourceReply.Create(messageBytes);
         else if (msgType == (short) MESSAGE_CLASS_IDS.PlayingFieldReply.getValue()) 
            result = ConfigurationReply.Create(messageBytes);
         else if (msgType == (short) MESSAGE_CLASS_IDS.AgentListReply.getValue()) 
            result = AgentListReply.Create(messageBytes);
         else if (msgType == (short) MESSAGE_CLASS_IDS.StatusReply.getValue()) 
            result = StatusReply.Create(messageBytes);
         else 
            throw new ApplicationException("Invalid Message Class Id", null);
        
        return result;
    }

    @Override
    public void Encode(ByteList messageBytes) throws UnknownHostException, NotActiveException, Exception {
        messageBytes.Add((short) MESSAGE_CLASS_IDS.Reply.getValue());                            // Write out this class id first
        messageBytes.update();

        short lengthPos = messageBytes.getCurrentWritePosition();    // Get the current write position, so we
        // can write the length here later
        messageBytes.Add((short) 0);                             // Write out a place holder for the length
        messageBytes.update();

        super.Encode(messageBytes);                              // Encode stuff from base class

        messageBytes.Add((byte) ReplyType.getValue());            // Write out a place holder for the length
        messageBytes.update();

        messageBytes.Add((byte) Status.getValue());               // Write out a place holder for the length
        messageBytes.update();

        if (Note == null) {
            Note = "";
        }
        messageBytes.Add(Note);
        messageBytes.update();
        short length = (short) (messageBytes.getCurrentWritePosition() - lengthPos - 2);
        messageBytes.WriteInt16To(lengthPos, length);           // Write out the length of this object        
    }

    @Override
    public void Decode(ByteList bytes) throws Exception {
        short objType = bytes.GetInt16();
        bytes.update();
        short objLength = bytes.GetInt16();
        bytes.update();

        bytes.SetNewReadLimit(objLength);
        bytes.update();

        super.Decode(bytes);
        bytes.update();

        ReplyType = PossibleTypes.convert((int) bytes.GetByte());
        bytes.update();
        Status = PossibleStatus.convert((int) bytes.GetByte());
        bytes.update();
        Note = bytes.GetString();

        bytes.RestorePreviosReadLimit();
    }

    public PossibleTypes getReplyType() {
        return ReplyType;
    }

    public void setReplyType(PossibleTypes replyType) {
        ReplyType = replyType;
    }

    public PossibleStatus getStatus() {
        return Status;
    }

    public void setStatus(PossibleStatus status) {
        Status = status;
    }

    public String getNote() {
        return Note;
    }

    public void setNote(String note) {
        Note = note;
    }

    @Override
    public short getClassId() {
        return (short) MESSAGE_CLASS_IDS.Reply.getValue();
    }

    public static int getMinimumEncodingLength() {
        MinimumEncodingLength = 4 // Object header
                + 1 // ReplyType
                + 1 // Status
                + 2;             // Note
        System.out.println("Reply.MinimumEncodingLength" + MinimumEncodingLength);
        return MinimumEncodingLength;
    }
}
